File: app/src/main/java/com/example/lenovo/reader/ReaderApplication.java
package com.example.lenovo.reader;

import android.app.Application;
import com.example.lenovo.reader.di.ApplicationComponent;
import com.example.lenovo.reader.di.ApplicationModule;
import org.jetbrains.annotations.NotNull;
import timber.log.Timber;

public class ReaderApplication extends Application {

    private ApplicationComponent applicationComponent;

    @Override
    public void onCreate() {
        super.onCreate();
        setUpTimber();
        setUpDependencyInjection();
    }

    private void setUpTimber() {
        Timber.plant(new Timber.DebugTree() {
            @Override
            protected String createStackElementTag(@NotNull StackTraceElement element) {
                return super.createStackElementTag(element) + ", " +
                        element.getClassName() + ", " +
                        element.getMethodName() + ", " +
                        element.getLineNumber();
            }
        });
    }

    private void setUpDependencyInjection() {
        applicationComponent = DaggerApplicationComponent.builder()
                .applicationModule(new ApplicationModule(this))
                //.modelModule(new ModelModule())
                .build();
    }

    private ApplicationComponent getApplicationComponent() {
        return applicationComponent;
    }
}

File: app/src/main/java/com/example/lenovo/reader/ReaderApplication.kt
package com.example.lenovo.reader

import com.example.lenovo.reader.controllers.DayNightController
import com.example.lenovo.reader.di.DaggerApplicationComponent
import com.example.model.bind
import com.example.model.schedule
import dagger.android.AndroidInjector
import dagger.android.DaggerApplication
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber
import javax.inject.Inject

class ReaderApplication : DaggerApplication() {

  @Inject
  lateinit var dayNightController: DayNightController

  var compositeDisposable: CompositeDisposable = CompositeDisposable()

  override fun onCreate() {
    super.onCreate()
    if (BuildConfig.DEBUG) {
      plantTimber()
    }
    dayNightController.applyPreferredDayNightMode()
      .schedule()
      .bind(compositeDisposable)
      .subscribe()
  }

  override fun applicationInjector(): AndroidInjector<out DaggerApplication> =
    DaggerApplicationComponent.builder().create(this)

  private fun plantTimber() {
    Timber.plant(object : Timber.DebugTree() {
      override fun createStackElementTag(element: StackTraceElement): String? =
        "${super.createStackElementTag(element)}, ${element.methodName}:: ${element.lineNumber}"
    })
  }

  override fun onTerminate() {
    super.onTerminate()
    compositeDisposable.dispose()
  }

}
